openprescribing/matrixstore/grouper.py

```diff
@@ -0,0 +1,56 @@
+from collections import defaultdict
+
+import numpy
+
+
+class Grouper(object):
+    """
+    The `Grouper` class provides a callable which sums together groups of
+    matrix rows. We generally use it for aggregating practice level data into
+    organisations that consist of groups of practices like CCGs or STPs.
+    """
```

Comment:
No, "grouper" is almost certainly not the right name but I'm not 100% sold on "summer" either, and not just because it's homographic for the season. The SQL operation it replaces is `SUM ... GROUP BY` and that's tricky to nounify. I wonder about calling the class `SumByGroup` and then calling instances of it 'sum_by_org' so we'd have code that looked like this:
```python
sum_by_org = sum_by(org_type)
items_matrix = sum_by_org(items_matrix)
```

Or maybe we could stick with "group by" in the variable name as it feels obvious to me that in this context we're summing within the groups:
```python
group_by_org = group_by(org_type)
items_matrix = group_by_org(items_matrix)
```

I don't think this is pure bikeshedding: it's worth getting right if we can.

Reply:
How about something like this:
```python
offsets_to_group = [(0, 'even'), (1, 'odd'), (2, 'even')]

matrix = np.array([
  [1, 2, 3, 4],
  [2, 3, 4, 5],
  [3, 4, 5, 6],
])

grouped = MatrixWithGroup(
    matrix,
    group_axis='rows',
    offsets_to_group=offsets_to_group)

# Included `output_groups` so it's explicit how the returned matrix is sorted row-wise
grouped.apply(np.sum, output_groups=['even', 'odd'])
```

Reply:
I wonder if it would be enough just to rename `Grouper` to `RowGrouper` and `__call__` to either `sum` or, if we want to be even more explicit, `sum_by_group`.

So the calling code would look like this:
```python
group_by_org = get_row_grouper(org_type)
items_matrix = group_by_org.sum(items_matrix)
```

Reply:
I quite like this.

Reply:
Yes, I think this is better and doesn't require changing much code.

I still quite like the idea of `group_by_org.apply(np.sum, items_matrix)` because it makes it very clear what's going on, but perhaps YAGNI in terms of any other aggregate function ever needing to be used.  And not particularly attached to it...